src/protocols/utils/dist.c

```diff
@@ -71,14 +71,13 @@ int nn_dist_send (struct nn_dist *self, struct nn_msg *msg,
     struct nn_dist_data *data;
     struct nn_msg copy;
 
-    /*  TODO: We can optimise for the case when there's only one outbound
+    /*  TODO: We can optimize for the case when there's only one outbound
         pipe here. No message copying is needed in such case. */
 
     /*  In the specific case when there are no outbound pipes. There's nowhere
-        to send the message to. Deallocate it. */
+        to send the message to. */
     if (nn_slow (self->count) == 0) {
-        nn_msg_term (msg);
-        return 0;
+        return -EAGAIN;
```

Comment:
I object to this change in principle, because you're trying to achieve something that makes no sense with pub/sub.

PUB/SUB is *best effort* delivery.  It doesn't tell you that the receiver(s) received it -- and indeed there could be some that receive it, and some that don't based on buffering state in the individual subscribers.

Basically, if you're expecting to be able to flood without dropping, then you can't use a multicast protocol like this on its own.  You can augment the protocol with a backchannel that is per subscriber to notice missed messages, and replay just those messages that are missed.

But if you need reliable delivery *DON'T USE PUB/SUB*.  The only SP protocol that is "reliable" is REQ/REP.

The "silent drop" behavior of the client here is a design *feature*, not a bug.  Think of this like radio broadcast -- you put the bits out on the wire (or maybe not, if the transmit queue is too full), but you have no idea if anyone was listening or received.

And frankly, if you try to fix *this* case, you'll quickly find that the other cases, where messages are dropped by individual pipes due to flow control for example, becomes an issue as well.  It's fundamentally not a solveable problem with just a single "multicast" channel.